### ISSUE TITLE
Configure Private Files By Default

### ISSUE THREAD

**Issue by OpenIT4u** [pantheon-systems/drops-8#69]
Hi, I've been trying the Pantheon Drupal 8 beta out and attempted to configure the use of Private files.

I've added the following line to settings.pantheon.php
`$settings['file_private_path']  = 'sites/default/files/private';`

I've created the private folder via SFTP.

The private file system path is recognised on `/admin/config/media/file-system`, but I cannot select Private as the default download method (the option is not there).

I get the following error message on various pages, but seems to be consistently at `/admin/config` which seems to be saying the the .htaccess file cannot be created.  This is regardless of the permissions on the private folder that I can set via SFTP.

```
Warning: file_exists(): Unable to find the wrapper "private" - did you forget to enable it when you configured PHP? in file_save_htaccess() (line 363 of core/includes/file.inc).
file_save_htaccess('private://', 1)
file_ensure_htaccess()
system_requirements('runtime')
call_user_func_array('system_requirements', Array)
Drupal\Core\Extension\ModuleHandler->invokeAll('requirements', Array)
Drupal\system\SystemManager->listRequirements()
Drupal\system\SystemManager->checkRequirements()
Drupal\system\Controller\SystemController->overview('system.admin_config')
call_user_func_array(Array, Array)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
call_user_func_array(Object, Array)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1)
Stack\StackedHttpKernel->handle(Object, 1, 1)
Drupal\Core\DrupalKernel->handle(Object)
```

It would be really good if the private file system worked out of the box as it does on the Pantheon Drupal 7 install.

**Comment by OpenIT4u** [comment]
Upgraded to beta 15.  

The private file system path is recognised and selectable on `/admin/config/media/file-system`.

I now get the following error on `/admin/config`:

```Notice: Undefined index: name in system_requirements() (line 45 of core/modules/system/system.install).
system_requirements('runtime')
call_user_func_array('system_requirements', Array)
Drupal\Core\Extension\ModuleHandler->invokeAll('requirements', Array)
Drupal\system\SystemManager->listRequirements()
Drupal\system\SystemManager->checkRequirements()
Drupal\system\Controller\SystemController->overview('system.admin_config')
call_user_func_array(Array, Array)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
call_user_func_array(Object, Array)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1)
Stack\StackedHttpKernel->handle(Object, 1, 1)
Drupal\Core\DrupalKernel->handle(Object)
Notice: Undefined index: version in system_requirements() (line 47 of core/modules/system/system.install).
system_requirements('runtime')
call_user_func_array('system_requirements', Array)
Drupal\Core\Extension\ModuleHandler->invokeAll('requirements', Array)
Drupal\system\SystemManager->listRequirements()
Drupal\system\SystemManager->checkRequirements()
Drupal\system\Controller\SystemController->overview('system.admin_config')
call_user_func_array(Array, Array)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array)
Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\{closure}()
call_user_func_array(Object, Array)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1)
Stack\StackedHttpKernel->handle(Object, 1, 1)
Drupal\Core\DrupalKernel->handle(Object)
One or more problems were detected with your Drupal installation. Check the status report for more information.
```
and the following in the dblog:
```Security warning: Couldn't write .htaccess file. Please create a .htaccess file in your <em class="placeholder">sites/default/files/config/active</em> directory which contains the following lines: <code><br /># Deny all requests from Apache 2.4+.<br /> &lt;IfModule mod_authz_core.c&gt;<br /> Require all denied<br /> &lt;/IfModule&gt;<br /> <br /> # Deny all requests from Apache 2.0-2.2.<br /> &lt;IfModule !mod_authz_core.c&gt;<br /> Deny from all<br /> &lt;/IfModule&gt;<br /> # Turn off all options we don&#039;t need.<br /> Options None<br /> Options +FollowSymLinks<br /> <br /> # Set the catch-all handler to prevent scripts from being executed.<br /> SetHandler Drupal_Security_Do_Not_Remove_See_SA_2006_006<br /> &lt;Files *&gt;<br /> # Override the handler again if we&#039;re run later in the evaluation list.<br /> SetHandler Drupal_Security_Do_Not_Remove_See_SA_2013_003<br /> &lt;/Files&gt;<br /> <br /> # If we know how to do it safely, disable the PHP engine entirely.<br /> &lt;IfModule mod_php5.c&gt;<br /> php_flag engine off<br /> &lt;/IfModule&gt;</code>
```

```Notice: Undefined index: version in system_requirements() (line 47 of /srv/bindings/02628db5b06846acac728f9a57e99d15/code/core/modules/system/system.install).
```

and

```Notice: Undefined index: name in system_requirements() (line 45 of /srv/bindings/02628db5b06846acac728f9a57e99d15/code/core/modules/system/system.install).
```

I will test if Private files is now working despite the above.

**Comment by OpenIT4u** [comment]
Following a brief test, private files appears to be working.  I changed the image field on the article content type to use the private download method.  I created an article with an image and saved it as unpublished.  Privileged users can see the article and file.  Logged off and could not access the image file directly.

Ideally Private files should work out of the box (no changes to settings.php) and without the error messages or security warnings.

**Comment by OpenIT4u** [comment]
I'm still getting the issue with a fresh Drupal 8 (beta 15) install.  I've updated settings.pantheon.php - the error messages are appearing again and whilst the private path is recognised, it's not accessible.

**Comment by greg-1-anderson** [comment]
Here is the change record: https://www.drupal.org/node/2392959

**Comment by greg-1-anderson** [comment]
Here is a blog on Private files for Drupal 7 sites:

https://pantheon.io/blog/privacy-please

Here is what I did to configure private files:

1. Switch to sftp mode.
2. "Open in your default SFTP client" from SFTP Connection Info
3. Create a new file "private" at sites/default/files/private
4. Get info on "default" folder and "settings.php" and make them writable.
5. Edit settings.php.
6. $settings['file_private_path']  = 'sites/default/files/private';
7. `drush @mysite.dev cr`

After completing these steps, "Private local files served by Drupal" became another selectable option under "Default download method" in admin/config/media/file-system?q=admin/config/media/file-system.

**Comment by greg-1-anderson** [comment]
![screen shot 2015-09-15 at 3 27 55 pm](https://cloud.githubusercontent.com/assets/612191/9891855/66773ece-5bbe-11e5-8008-d7a4bca0c07e.png)

**Comment by OpenIT4u** [comment]
Hi Greg

Those are the steps I pretty much worked out - with the exception of
clearing the cache.  However I fired up a D7 instance and am pretty sure I
did not need to - the option to chose between public and private was there
as default.  I'll double check later.  As long as D8 is no worse to set up
than D7 then that's fine.

In the meantime this can probably be downgraded from Major.

Matt Barton
*OpenIT4u.com* <http://OpenIT4u.com>

Open Source Technology - Green Hosting - Drupal Expertise

**Comment by greg-1-anderson** [comment]
@OpenIT4u This is the 'drops-8' queue, so Drupal 7 is out of scope here. In my experience, configuring the private files location in settings.php was sufficient to make the "Private file system path", but it was necessary to clear the cache before you could select it as the default download location.

We might improve the experience here by pre-populating the private files setting in settings.php, so that private files are an option from the get-go for sites created on Pantheon.

This is labeled "major" because it is a required documentation task that must be completed before Pantheon officially supports Drupal 8.

**Comment by bmackinney** [comment]
@greg-1-anderson the docs issue was completed: https://github.com/pantheon-systems/documentation/issues/937 Can this issue be closed?

**Comment by greg-1-anderson** [comment]
I still need to configure the private files setting in settings.panthon.inc.
